### PR TITLE
Split large rec files between multiple SpikegadgetsRawIO iterators

### PR DESCRIPTION
**Pull request by samuelbray32**
Excessively large iterators can cause memory issues due to lading large chunks of data into memory during `get_analogsignal_multiplexed` or `get_digitalsignal`. 

This merge splits the memmap of these large rec files into multiple SpikegadgetsRawIO iterators.  The ordering of these is tracked by the `RecDataChunkIterator` so almost any use case doesn't need to worry about this.

If trying too get info for a specific epoch, note that this means there is not a 1:1 relation between iterators and epochs.  Use the `file_info` pandas dataframe to determine which epoch each belongs to (see `add_epochs` for example)

### CODE REVIEW

File: src/spikegadgets_to_nwb/convert_ephys.py [hunk 1]
```diff
@@ -83,6 +84,43 @@ def __init__(
...
+                            i_start=partial_size - 10,
+                            i_stop=partial_size,
+                            padding=30000,
+                        )[
+                            -1
+                        ]
+                    j += MAXIMUM_ITERATOR_SIZE
+                self.neo_io.pop(iterator_loc)
+                self.neo_io[iterator_loc:iterator_loc] = sub_iterators
+        print("# iterators:", len(self.neo_io))
```
Comment (edeno) [inline src/spikegadgets_to_nwb/convert_ephys.py:123]: Should this use the logger?

File: src/spikegadgets_to_nwb/spike_gadgets_raw_io.py [hunk 2]
```diff
@@ -537,6 +537,93 @@ def get_analogsignal_multiplexed(self, channel_names=None) -> np.ndarray:
...
+            interleaved_data_id_byte_values >> data_offsets[:, 2]
+        ) & 1
+        # calculate which packets encode for which channel
+        initialize_stream_mask = np.logical_or(
+            (np.arange(num_packet) == 0)[:, None], interleaved_data_id_bit_values == 1
+        )
+        # read the data into int16
+        data = (
+            self._raw_memmap[i_start:i_stop, data_offsets[:, 0]]
+            + self._raw_memmap[i_start:i_stop, data_offsets[:, 0] + 1] * 256
```
Comment (edeno) [inline src/spikegadgets_to_nwb/spike_gadgets_raw_io.py:615]: Where does this 256 come from? int16?

**Review by edeno — APPROVED** [review]

File: src/spikegadgets_to_nwb/spike_gadgets_raw_io.py (hunk 2 above)
Comment (samuelbray32) [reply src/spikegadgets_to_nwb/spike_gadgets_raw_io.py:615]: Yeah, this was a way to vectorize the conversion

Comment (edeno) [reply src/spikegadgets_to_nwb/spike_gadgets_raw_io.py:615]: Can you make this an informative constant just so there are no magic variables.

File: src/spikegadgets_to_nwb/convert_ephys.py [hunk 3]
```diff
@@ -83,6 +84,43 @@ def __init__(
...
+                        )
+                    )
+                    if self.n_multiplexed_channel > 0:
+                        partial_size = sub_iterators[-1]._raw_memmap.shape[0]
+                        previous_multiplex_state = sub_iterators[
+                            -1
+                        ].get_analogsignal_multiplexed_partial(
+                            i_start=partial_size - 10,
+                            i_stop=partial_size,
+                            padding=30000,
```
Comment (edeno) [inline src/spikegadgets_to_nwb/convert_ephys.py:116]: Is this padding related to the sampling rate? What if there is a different sampling rate in the data?

Comment (edeno) [reply src/spikegadgets_to_nwb/convert_ephys.py:116]: For example, there are some datasets with 20 kHz sampling rate

File: src/spikegadgets_to_nwb/spike_gadgets_raw_io.py [hunk 4]
```diff
@@ -713,3 +800,143 @@ def access_coordinates(self, index):
...
+            interleaved_data_id_byte_values >> data_offsets[:, 2]
+        ) & 1
+        # calculate which packets encode for which channel
+        initialize_stream_mask = np.logical_or(
+            (np.arange(num_packet) == 0)[:, None], interleaved_data_id_bit_values == 1
+        )
+        # read the data into int16
+        data = (
+            self._raw_memmap[:, data_offsets[:, 0]]
+            + self._raw_memmap[:, data_offsets[:, 0] + 1] * 256
```
Comment (edeno) [inline src/spikegadgets_to_nwb/spike_gadgets_raw_io.py:923]: Probably want to use the constant here as well

File: src/spikegadgets_to_nwb/spike_gadgets_raw_io.py [hunk 5]
```diff
@@ -537,6 +537,93 @@ def get_analogsignal_multiplexed(self, channel_names=None) -> np.ndarray:
             )
         return analog_multiplexed_data
 
+    def get_analogsignal_multiplexed_partial(
+        self,
+        i_start: int,
+        i_stop: int,
+        channel_names: list = None,
+        padding: int = 30000,
```
Comment (edeno) [inline src/spikegadgets_to_nwb/spike_gadgets_raw_io.py:547]: Is this also related to sampling rate?

**Review by edeno — APPROVED** [review]
